### DISCUSSION TITLE
Provide a generic way to `delay` 3rd-party authored methods/function

### DISCUSSION THREAD

**Discussion by alanjds** [celery/celery#7048] [Ideas]
# Checklist

- [x] I have checked the [issues list](https://github.com/celery/celery/issues?q=is%3Aissue+label%3A%22Issue+Type%3A+Enhancement%22+-label%3A%22Category%3A+Documentation%22)
  for similar or identical enhancement to an existing feature.
- [x] I have checked the [pull requests list](https://github.com/celery/celery/pulls?q=is%3Apr+label%3A%22Issue+Type%3A+Enhancement%22+-label%3A%22Category%3A+Documentation%22)
  for existing proposed enhancements.
- [ ] I have checked the [commit log](https://github.com/celery/celery/commits/master)
  to find out if the if the same enhancement was already implemented in the
  master branch.
- [x] I have included all related issues and possible duplicate issues in this issue
      (If there are none, check this box anyway).

#### Related Issues

- None

#### Possible Duplicates

- None

# Brief Summary

This proposes Celery package to provide a generic `delay(func, *args, **kwargs)` that wraps any func or method thrown at it.

# Design

Historically, to make a function "able" to be delayed via Celery you need to wrap it in a decorator at least. This can be ok for own code, but not for 3rd-party code. Also makes your code coupled with Celery or at least with Celery API for jobs.

"Modern" Python versions can natively pickle functions, so this is no more needed for simpler uses. If Celery provides a shared task that receives a function and its arguments, no custom code is needed for standard user code.

Example:

```py
# mymodule.py
import requests
def ping_website(url):
    requests.get(url)

# main.py
from celery.contrib import delay

from .mymodule import ping_website
from .myothermodule import greet

delay(ping_website, 'http://www.google.com.br')
delay(greet)
```

Note how the Celery dependent part occur only on the main module, not on each of function definitions. Even better, as we do not care about the response, such call could be wrote as:

```py
delay(requests.get, 'http://www.google.com.br')
```

This is not possible with today's prescription of "create a Task to be able to .delay it", you need to import the 3rd-party funcs and code in a task that wraps it to be able to send to Celery.  

## Architectural Considerations

Code `delay()`ed can be considered more convenient. Yet could potentially carry more data than needed or more data than the former way of calling. People can optimize by writing in the former way if needed.

Methods carry the whole `self` when serialized and this should be informed to the user, as they can consider if is acceptable to pay the cost to do so. Also is needed to inform that such feature needs pickle and reproduce the notice about the danger of accepting pickled jobs.

## Proposed Behavior

I propose some code along the following to be included on Celery release and the usage to be featured on Quickstart as the simpler way to sideload your code via Celery

Code to be included:
```py
# celery/contrib.py
@shared_task(serializer='pickle')
def _call(func, *args, **kwargs):
    return func(*args, **kwargs)


def delay(func, *args, **kwargs) -> celery.result.AsyncResult:
    """
    Defer a call of func(*args, **kwargs) to a Celery worker
    """
    return _call.delay(func, *args, **kwargs)
```

Example to be included in docs:
```py
from celery.contrib import delay

# synchronous version:
my_slow_procedure(3, 4, spam='eggs')

# sideloaded version:
delay(my_slow_procedure, 3, 4, spam='eggs')
``` 

## Proposed UI/UX

The simpler usage is to `delay(fn, *args, **kwargs)` arbitrary functions/methods.

Anything more advanced than this could be considered advanced enough to honor the today's way of do it.

## Diagrams

N/A

## Alternatives

None

**Comment by auvipy** [comment]
can you contribute to this?

**Comment by alanjds** [comment]
> can you contribute to this?

Sure.

Opened as an issue to gather some feedback on the proposal.

Can I consider it as accepted in design?

**Comment by auvipy** [comment]
> > can you contribute to this?
> 
> Sure.
> 
> Opened as an issue to gather some feedback on the proposal.
> 
> Can I consider it as accepted in design?

it seems OK as per your description to come with a Proof of concept. yet some more thoughts could be shared when the implementation can be seen. I am personally OK to give the POC a try. @celery/technical-board & @celery/core-developers will also chime in where necessary. But in general please go for it as it seems a useful influsion.
